crates/bevy_ecs/src/query/state.rs

```diff
@@ -508,22 +508,26 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
         archetype: &Archetype,
         access: &mut Access<ArchetypeComponentId>,
     ) {
-        self.component_access
-            .access
-            .component_reads()
-            .for_each(|id| {
-                if let Some(id) = archetype.get_archetype_component_id(id) {
-                    access.add_component_read(id);
-                }
-            });
-        self.component_access
-            .access
-            .component_writes()
-            .for_each(|id| {
-                if let Some(id) = archetype.get_archetype_component_id(id) {
-                    access.add_component_write(id);
-                }
-            });
+        for component_id in archetype.components() {
```

Comment:
I think we expect archetypes to usually have more components than queries, so flipping the iteration direction unconditionally could be a performance regression.  It might be worth doing 

```rust
if self.component_access.access.component_reads_and_writes_inverted() {
    // New code
} else {
    // Old code
}
```

But I haven't profiled it, so it may not matter at all.